executor/src/witgen/affine_expression.rs

```diff
@@ -444,28 +421,110 @@ where
+    fn add(self, rhs: Self) -> Self::Output {
+        use std::cmp::Ordering;
+        use AffineExpression::*;
+        match (self, rhs) {
+            // constant + constant
+            (Constant(o1), Constant(o2)) => Constant(o1 + o2),
+            // constant + one
+            (One(coeff, o1), Constant(o2)) | (Constant(o2), One(coeff, o1)) => One(coeff, o1 + o2),
+            // constant + many
+            (
+                Many {
+                    coefficients,
+                    offset,
+                },
+                Constant(o),
+            )
+            | (
+                Constant(o),
+                Many {
+                    coefficients,
+                    offset,
+                },
+            ) => Many {
+                coefficients,
+                offset: offset + o,
+            },
+            // one + one
+            (One((v1, c1), o1), One((v2, c2), o2)) => match v1.cmp(&v2) {
+                Ordering::Equal => {
+                    let c = c1 + c2;
+                    if c.is_zero() {
+                        Constant(o1 + o2)
+                    } else {
+                        One((v1, c), o1 + o2)
+                    }
+                }
+                Ordering::Less => Many {
+                    coefficients: vec![(v1, c1), (v2, c2)],
+                    offset: o1 + o2,
+                },
+                Ordering::Greater => Many {
+                    coefficients: vec![(v2, c2), (v1, c1)],
+                    offset: o1 + o2,
+                },
+            },
+            // one + many
+            (
+                One((v, c), o),
+                Many {
+                    mut coefficients,
+                    offset,
+                },
+            )
+            | (
+                Many {
+                    mut coefficients,
+                    offset,
+                },
+                One((v, c), o),
+            ) => {
+                debug_assert!(Self::is_proper_coefficients(&coefficients));
+                match coefficients.binary_search_by_key(&&v, |(k, _)| k) {
+                    Ok(idx) => coefficients[idx].1 += c,
+                    Err(idx) => coefficients.insert(idx, (v, c)),
+                }
+                coefficients.retain(|(_, c)| !c.is_zero());
+                if coefficients.is_empty() {
+                    Constant(offset + o)
+                } else {
+                    Many {
+                        coefficients,
+                        offset: offset + o,
+                    }
+                }
+            }
+            // many + many
+            (
+                Many {
+                    coefficients: mut c1,
+                    offset: o1,
+                },
+                Many {
+                    coefficients: c2,
+                    offset: o2,
+                },
+            ) => {
+                debug_assert!(Self::is_proper_coefficients(&c1));
+                for (k2, v2) in c2 {
```

Comment:
This could be faster if you used some kind of merge sort here. Also you could swap `c1` and `c2` if `c2` is larger.

Reply:
changed it to a parallel iteration + merge and it is indeed a bit faster (~3%) for the many chunks benchmark!